OrbitCore/CallstackData.h

```diff
@@ -32,63 +32,61 @@ class CallstackData {
   void AddCallStackFromKnownCallstackData(const orbit_client_protos::CallstackEvent& event,
                                           const CallstackData* known_callstack_data);
 
-  [[nodiscard]] const BlockChain<orbit_client_protos::CallstackEvent, 16 * 1024>& callstack_events()
-      const {
-    return callstack_events_;
-  };
-
-  [[nodiscard]] uint32_t GetCallstackEventsSize() const { return callstack_events_.size(); };
-
-  [[nodiscard]] const CallStack* GetCallStack(CallstackID callstack_id) const {
-    absl::MutexLock lock(&unique_callstacks_mutex_);
-    auto it = unique_callstacks_.find(callstack_id);
-    if (it != unique_callstacks_.end()) {
-      return it->second.get();
-    }
-    return nullptr;
-  };
-
-  [[nodiscard]] bool HasCallStack(CallstackID callstack_id) const {
-    absl::MutexLock lock(&unique_callstacks_mutex_);
-    return unique_callstacks_.count(callstack_id) > 0;
+  [[nodiscard]] const absl::flat_hash_map<int32_t,
+                                          std::map<uint64_t, orbit_client_protos::CallstackEvent>>&
+  callstack_events_by_tid() const {
+    return callstack_events_by_tid_;
   }
 
-  void ForEachUniqueCallstack(const std::function<void(const CallStack&)>& action) const {
-    absl::MutexLock lock(&unique_callstacks_mutex_);
-    for (const auto& it : unique_callstacks_) {
-      action(*it.second);
-    }
-  }
+  [[nodiscard]] uint32_t GetCallstackEventsCount() const;
+
+  [[nodiscard]] std::vector<orbit_client_protos::CallstackEvent> GetCallstackEventsInTimeRange(
+      uint64_t time_begin, uint64_t time_end) const;
+
+  [[nodiscard]] absl::flat_hash_map<int32_t, uint32_t> GetCallstackEventsCountsPerTid() const;
+
+  [[nodiscard]] uint32_t GetCallstackEventsOfTidCount(int32_t thread_id) const;
+
+  [[nodiscard]] std::vector<orbit_client_protos::CallstackEvent> GetCallstackEventsOfTidInTimeRange(
+      int32_t tid, uint64_t time_begin, uint64_t time_end) const;
+
+  void ForEachCallstackEvent(
+      const std::function<void(const orbit_client_protos::CallstackEvent&)>& action) const;
+
+  void ForEachCallstackEventOfTid(
+      int32_t tid,
+      const std::function<void(const orbit_client_protos::CallstackEvent&)>& action) const;
+
+  [[nodiscard]] uint64_t max_time() const { return max_time_; }
```

Comment:
I would protect this by mutex too, since you are modifying them under mutex.

Reply:
Done.